### ISSUE TITLE
Better support for categorical predictors

### ISSUE THREAD

**Issue by Fish-Soup** [scikit-learn-contrib/py-earth#156]
Hi i am also interested in the above. I had an idea, not sure how hard it would be to implement. I have a good example.

Basically I am trying to model gas and electricity consumption. Now its fairly obvious that this depends on businesses working weeks. So in this case we would have 7 (0,1) flags for the day of the week. However in all likely hood what i expect to dominate is a weekend weekday effect. If i give the 7 columns with day of week flags i expect that MARS would find it harder to find a weekday weekend effect as it would have to use 7 variables to do this. Where as if i where to simply give a weekday/weekend flag as this would only require 1 variable, MARS would find this less 'costly' to fit and is more likely to find a match.

But in the above case where i provide 7 day of week flags plus a weekend/weekday flag i am essentially choosing how to split the week up, where as i would hope MARS could do this itself. 

I wonder if MARS could derive from the 7 day of week flags that days 1-5 are the same and 6-7 the same, and create a variable that is less costly within the algorithm to fit.

1) When passed a xdata entry of strings it makes a set of flag columns 1 for each categorical data point (this would make things more convenient and allow the 2) ) 
2) Once MARS has done the initial over fit it sees if the the parameters fitted for say  Monday to Friday are essentially the same and as such can be grouped. the grouped version creates a new variable, which is added tot he list of variables. Now MARS is refitted with weekday, weekend as well as the 7 day of week flags. Thus if we don't get a good enough fit p[er complexity of the 7 day of week flags, the weekend weekday effect can dominate.  Once grouped the complexity cost of fitting them is reduced. 

I suppose with 2) you could use brute force and find every combination of the 7 day of week flags but i expect this will create a massive amount of categorical flags: using N choose R i get 127 flags from 7.

The above example is kinda obvious and i can work around it, but there are other cases where say i am forecasting 100 counter parties. Say some are offices and some are factories, if I had categorical flag for each site then i would like MARS to find the groups of sites. say one set are factories and one set are offices. 

This could well be beyond scope or take too long for MARS to fit but i thought i would mention it if you where thinking about modifying categorical predictors.

Cheers

**Comment by jcrudy** [comment]
@Fish-Soup, there is some existing work on MARS for categorical variables.  See here:

http://media.salford-systems.com/library/MARS_V2_JHF_LCS-108.pdf

The discussion of categorical variables starts on page 13.  It seems quite similar to what you suggest.  My initial plan would be to implement the technique described.  However, I'm not sure how soon it will happen.  If you feel like reading it, I would be interested to learn your thoughts on whether the proposed techniques would meet your needs.  The paper is, unfortunately, rather wordy, though.

**Comment by empirical-bayesian** [comment]
Not wanting to be a stick in the mud, but series with categorical
predictors can also be modeled by more traditional means, including state-
space methods developed and taught by people like Durbin, Koopman,
Harvey, Poole, West, and Harrison. If time -- basically an index of
successive dependent observations -- is less important, then the
generalized additive models that Wood so ably describes[1], even if they
are frequentist, warrant a look. Because I'm a Bayesian I must caution,
however, that these are *so* flexible, it's possible these are
essentially model-free, unless strong priors are applied to components,
like the [2]*g-priors[3]* of Bové, Held, and Kauermann[4].
Earth/MARS might produce low m.s.e., but it can be hard to constrain
model elements, and harder to interpret such constraints in problem-
specific terms.
 - Jan Galkowski

On Wed, Apr 26, 2017, at 05:26, Fish-Soup wrote:
> Hi i am also interested in the above. I had an idea, not sure how hard
> it would be to implement. I have a good example.> Basically I am trying to model gas and electricity consumption. Now
> its fairly obvious that this depends on businesses working weeks. So
> in this case we would have 7 (0,1) flags for the day of the week.
> However in all likely hood what i expect to dominate is a weekend
> weekday effect. If i give the 7 columns with day of week flags i
> expect that MARS would find it harder to find a weekday weekend effect
> as it would have to use 7 variables to do this. Where as if i where to
> simply give a weekday/weekend flag as this would only require 1
> variable, MARS would find this less 'costly' to fit and is more likely
> to find a match.> But in the above case where i provide 7 day of week flags plus a
> weekend/weekday flag i am essentially choosing how to split the week
> up, where as i would hope MARS could do this itself.> I wonder if MARS could derive from the 7 day of week flags that days
> 1-5 are the same and 6-7 the same, and create a variable that is less
> costly within the algorithm to fit.


>  1. When passed a xdata entry of strings it makes a set of flag
>     columns 1 for each categorical data point (this would make things
>     more convenient and allow the 2) )
>  2. Once MARS has done the initial over fit it sees if the the
>     parameters fitted for say  Monday to Friday are essentially the
>     same and as such can be grouped. the grouped version creates a new
>     variable, which is added tot he list of variables. Now MARS is
>     refitted with weekday, weekend as well as the 7 day of week flags.
>     Thus if we don't get a good enough fit p[er complexity of the 7
>     day of week flags, the weekend weekday effect can dominate.  Once
>     grouped the complexity cost of fitting them is reduced.> I suppose with 2) you could use brute force and find every combination
> of the 7 day of week flags but i expect this will create a massive
> amount of categorical flags: using N choose R i get 127 flags from 7.> The above example is kinda obvious and i can work around it, but there
> are other cases where say i am forecasting 100 counter parties. Say
> some are offices and some are factories, if I had categorical flag for
> each site then i would like MARS to find the groups of sites. say one
> set are factories and one set are offices.> This could well be beyond scope or take too long for MARS to fit but i
> thought i would mention it if you where thinking about modifying
> categorical predictors.> Cheers


> — You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub[5], or mute the
> thread[6].> 


Links:

  1. https://www.crcpress.com/Generalized-Additive-Models-An-Introduction-with-R-Second-Edition/Wood/p/book/9781498728331
  2. https://arxiv.org/pdf/1108.3520.pdf
  3. https://arxiv.org/pdf/1108.3520.pdf
  4. https://arxiv.org/pdf/1108.3520.pdf
  5. https://github.com/scikit-learn-contrib/py-earth/issues/156

**Comment by Fish-Soup** [comment]
@jcrudy I had a quick read of the document. It seems to be in line with what I was thinking. I wonder if something aproaching this could also be applied to continuous data where by you could split two region's into non continous chunck's.  In my example sometimes with time  factories turn on and start consuming on say a specific day of the year.  MARS normally acts a bit idly around this jump in demand a's the underlying distribution is not smooth but Mars tries to make it so.  

@empirical-bayesian I will look into gam's currently I have a mixture of continous and catagoricall data. Overall MARS is fitting it well.

**Comment by jcrudy** [comment]
@Fish-Soup It would certainly be possible to add something like that.  For now, though, I think the trick from our issue #141 discussion is the closest that you'll get.  If I do get around to implementing the subsetting approach described in that paper, though, perhaps in the process I'll think about this as well.

@empirical-bayesian I would definitely agree that with only categorical variables the case for using MARS is not that strong.  Although, with the automatic subset selection process described in that paper it might be stronger.  The main advantage of MARS over GAM, as I've seen them implemented (and I'm no expert on this), is the automatic search for interactions.  However, I believe GAM has its own advantages, such as a greater variety of loss functions.  I'm not sure what functionality GAM offers for missing data.

**Comment by empirical-bayesian** [comment]
I haven't dug into *py-earth*'s implementation for multiple interactions
and especially automatic search.  I am leery, however, because if a
significant interaction is tested using a traditional hypothesis test,
then seeking multiple ones needs correction for multiple testing, per
*familywise error rates* and the like. (See Efron, *Large-Scale
Inference*, 2010, and, in particular, *false discovery rate control*.)
Interactions are also tough to validate ... If there's a particular high
order combination of attribute predictors which enter an interaction,
the number of observations (responses bound to predictors) for each such
interaction are smaller, possibly *much* smaller. Moreover the number of
observations across interactions is not likely to be *balanced*, unless
the experimental design is done to assure this.
 - Jan

On Thu, May 11, 2017, at 14:09, Jason Rudy wrote:
> @Fish-Soup[1] It would certainly be possible to add something like
> that.  For now, though, I think the trick from our issue #141[2]
> discussion is the closest that you'll get.  If I do get around to
> implementing the subsetting approach described in that paper, though,
> perhaps in the process I'll think about this as well.> @empirical-bayesian[3] I would definitely agree that with only
> categorical variables the case for using MARS is not that strong.
> Although, with the automatic subset selection process described in
> that paper it might be stronger.  The main advantage of MARS over GAM,
> as I've seen them implemented (and I'm no expert on this), is the
> automatic search for interactions.  However, I believe GAM has its own
> advantages, such as a greater variety of loss functions.  I'm not sure
> what functionality GAM offers for missing data.> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[4], or mute the thread[5].> 


Links:

  1. https://github.com/fish-soup
  2. https://github.com/scikit-learn-contrib/py-earth/issues/141
  3. https://github.com/empirical-bayesian
  4. https://github.com/scikit-learn-contrib/py-earth/issues/156#issuecomment-300871938

**Comment by Fish-Soup** [comment]
@jcrudy Hi yeah i still mean to try the approach in #141 , Other stuff just keeps coming up that I need to solve first. And yeah I mainly mentioned my idea in case you did get to implementing the categorical data improvements as it did seem like an consistent addition.

@empirical-bayesian For my current simulations the interaction terms are very important. I'm sure its possible to model them with more care in other approaches. But currently I need a method to apply to a few different problems as their delivery is time critical. Once i have decent solutions across the board I will look into other approaches. I just currently find MARS quite easy to use and understand the output.

**Comment by jcrudy** [comment]
Definitely agreed about the dangers of considering many possible interactions.  MARS does make use of various methods to reduce the risk of overfitting, but I would always recommend cross-validation with MARS or any other machine learning technique.
